cupy/_core/_kernel.pyx

```diff
@@ -1618,20 +1673,49 @@ cdef class _Ops:
                         (dtype, name))
 
     cpdef _Op _guess_routine_from_in_types(
-            self, tuple in_types, object can_cast=_numpy_can_cast):
+            self, tuple in_types, tuple weaks, object can_cast=_numpy_can_cast
+    ):
         cdef _Op op
         cdef tuple op_types
         cdef Py_ssize_t n = len(in_types)
         cdef Py_ssize_t i
+
+        in_categories = [-1 if weaks[i] else _get_kind_score2(in_types[i])
+                         for i in range(n)]
+        max_category = max(in_categories)
+
+        in_prec = [-1 if weaks[i] else _get_precision_score(in_types[i])
+                   for i in range(n)]
+        max_prec = max(in_prec)
+
+        if max_category == -1:
+            # all in_args are scalars
+            _weaks = (False,)*n
+        else:
+            _weaks = list(weaks)
+            for i in range(n):
+                if weaks[i] and _get_kind_score2(in_types[i]) > max_category:
+                    _weaks[i] = False
+                else:
+                    _weaks[i] = weaks[i]
+
         for op in self.ops:
             op_types = op.in_types
             for i in range(n):
                 it = in_types[i]
                 ot = op_types[i]
+                is_weak = _weaks[i]
+
+                # special case: f32 array + 1j -> c64 (not c128)
+                if weaks[i] and it == numpy.complex128 and max_prec == 2:
+                    it = numpy.complex64
+                    is_weak = False
+
                 if isinstance(it, tuple):
+                    # XXX: account for weaks
                     if not can_cast(it[0], ot) and not can_cast(it[1], ot):
                         break
-                elif not can_cast(it, ot):
+                elif not is_weak and not can_cast(it, ot):
```

Comment:
I know it is a bit slow, but it should be cached, so I would suggest replacing the full above logic logic with:
```
try:
    safe = np.result_type(weak(0), ot) == ot
except TypeError:
    safe = False
```
Might not be the prettiest ever, but I think it is preferable to custom implementation of the category logic and more flexible for future additions.

I am suggesting another trick:  Let's make `weaks = [False, int, float, complex]` i.e. store either `False` _or_ the type, so that we can use the type information directly from there?

Reply:
Currently, `in_types` entries are numpy scalars types already, after the call to https://github.com/cupy/cupy/blob/622cdf4ae126d8bf611eeb0e0dd77a5e069b5cd5/cupy/_core/_kernel.pyx#L180 at https://github.com/cupy/cupy/blob/622cdf4ae126d8bf611eeb0e0dd77a5e069b5cd5/cupy/_core/_kernel.pyx#L1313

Which is the main reason to have an additional `weaks` tuple to be able to distinguish `int8(1) + 1` from `int8(1) + int8(1)`.

So your suggestion is to keep python types to be python types all the way instead of converting them to numpy types early on?